contracts/Staking.sol

```diff
@@ -458,66 +446,110 @@ contract Staking is Autopetrified, ERCStaking, ERCStakingHistory, IStakingLockin
         totalStakedHistory.add64(getBlockNumber64(), newStake);
     }
 
-    /**
-     * Note: So far this function is called from unlock and from transferFromLock,
-     * which both ensure that lock can be unlocked (the latter through isLockManager modifier)
-     */
-    function _unlock(address _accountAddress, uint256 _lockId) internal {
+    function _unsafeUnlock(address _accountAddress, address _lockManager) internal {
         Account storage account = accounts[_accountAddress];
-        Lock storage lock_ = account.locks[_lockId];
+        Lock storage lock_ = account.locks[_lockManager];
 
-        lock_.unlockedAt = getTimestamp64();
+        // update total
+        // no need for SafeMath: totalLocked must be greater or equal than lock._amount
+        account.totalLocked = account.totalLocked - lock_.amount;
 
-        // remove from active locks, replacing it by the last one in the array
-        // we assume consistency here, i.e., that lock exists in active array
-        uint256 locksLength = account.activeLockIds.length;
-        if (locksLength == 1) {
-            delete account.activeLockIds;
-            return;
-        }
+        emit Unlocked(_accountAddress, _lockManager, lock_.amount);
 
-        for (uint256 i = 0; i < locksLength; i++) {
-            if (account.activeLockIds[i] == _lockId) {
-                account.activeLockIds[i] = account.activeLockIds[locksLength - 1];
-                delete account.activeLockIds[locksLength - 1];
-                account.activeLockIds.length--;
-                break;
-            }
-        }
+        delete account.locks[_lockManager];
+    }
 
-        emit Unlocked(_accountAddress, _lockId, lock_.amount, lock_.manager, lock_.data);
+    function _increaseLockAllowance(address _lockManager, Lock storage _lock, uint256 _allowance) internal {
+        require(_allowance > 0, ERROR_AMOUNT_ZERO);
+
+        _lock.allowance = _lock.allowance.add(_allowance);
+
+        emit LockAllowanceChanged(msg.sender, _lockManager, _allowance, true);
     }
 
-    function _updateActiveLockAmount(address _accountAddress, uint256 _lockId, uint256 _amount, bool _increase) internal {
-        Lock storage lock_ = accounts[_accountAddress].locks[_lockId];
-        // check that lock hasn't been unlocked
-        require(lock_.unlockedAt > getTimestamp64(), ERROR_UNLOCKED_LOCK); // locks are created with a MAX_UINT64 unlockedAt
-        // checking that lock is in active array shouldn't be needed if data is consistent
+    function _increaseActiveLockAmount(address _accountAddress, address _lockManager, uint256 _amount) internal {
```

Comment:
Not sure if I understand correctly could we have an "inactive" lock?

Reply:
Well, I consider a lock active when it has positive allowance. If allowance is zero, you could consider that the lock doesn’t exist, but the object will, as it’s a mapping.
So it’s a way to emphasize that you need to set allowance first in order to use this function.